### ISSUE TITLE
DataGrid Pagination when rows are grouped: Option to paginate by rows instead of groups

### ISSUE THREAD

**Issue by matlangenberger** [mui/mui-x#10417]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Summary 💡

When a rowGroupModel is defined, there should be an option to have pagination use 'rows' instead of 'groups' in determining how many items are on a given page as well as the 'total rows' listed in the pagination component. It should be possible to set rowsPerPage to be equal to a 'row count' instead of a 'group count' when row grouping is enabled. If the group traverses pages, it should just continue onto the next page with the rowGroup showing again at the top of the second page.

### Examples 🌈

For example i have roughly 15k records grouped in a 3 tier rowGroupModel. Some of these groups are very large while others are very small. I want to be able to paginate by 50 rows (for example) instead of 3 groups which can vary largely in size. The current result is that I have some pages with hundreds of rows and others with maybe 15. i would like to have these all be the same size regardless of how large the groups are.

**Comment by cherniavskii** [comment]
Hi @matlangenberger 
This is the limitation of our current pagination implementation.

Wouldn't this behavior be misleading though?
When the row group is split into multiple pages, it might not be obvious why the same group appears multiple times (on different pages).

**Comment by matlangenberger** [comment]
@cherniavskii thanks for the quick response.

I actually find that the current presentation is counterintuitive for my particular use case.

I have roughly 400 client groups with about 15,000 rows in total. I have a handful of really large client groups that contain hundreds of thousands of rows and some client groups that have 3 or 4 rows. The result is a really wonky experience where some pages are infinitely long and some don't even take up a full page especially when filtered. It's also difficult for the user to see how many rows are present on the screen since the number of groups is displayed in the pagination component instead of rows. 

I think having the option to paginate by rows would alleviate that and I think the group continuing on the next page could be clarified by having some type of suffix added like (cont...) maybe with the ability to disable or specify an alternative suffix.

I had this exactly proposed implementation using another table library and it was clear when groups overflowed onto a new page.

I appreciate you considering my proposal.

**Comment by flash548** [comment]
@cherniavskii  @matlangenberger Yeah I have some similar scenarios that would involve something where an expanded parent row might contain a number of rows that would exceed the grid's current pagination model `pageSize`, but I'm kind of with the maintainers here in how that would even be done... if you're several pages deep into an expanded grouping (of an expanded parent); how would you know what parent node you even expanded 5 pages ago?  And now say you want to collapse that giant list, do you go back to page 1 where the expansion was - and that's assuming its on page 1.

But say pagination of data rows in this scenario was even possible, now throw in multiple groupings, say the 2nd or 3rd level of expansion for a given grouped row yields a huge list of subgroups that exceeds your expected/set pagination size. Now you wish you could paginate on this giant list of subgroups, and not the data rows.

The need is there; but I'm trying to think how this would even work. Because you might well likely have cases where you'd need _both_ depending on your data and use-case, and I don't know how a feature like that would be implemented.  Overall I think any path would lead to a strange UI/UX experience.
